src/multisig.cpp

```diff
@@ -0,0 +1,402 @@
+            std::vector<Crypto::SecretKey> keys = dedupeKeys(secretKeys);
+
+            /* Sort the keys in order as it is important that they
+               are added together in order */
+            sortKeys(keys);
+
+            sharedSecretKey = addKeys(keys);
+        }
+
+        /* Helper method that calculates a shared private key by adding the vector
+           of private keys together and returing the result directly */
+        Crypto::SecretKey calculate_shared_private_key(const std::vector<Crypto::SecretKey> &secretKeys)
+        {
+            Crypto::SecretKey secretKey;
+
+            calculate_shared_private_key(secretKeys, secretKey);
+
+            return secretKey;
+        }
+
+        /* Calculates a shared public key by adding the vector
+           of public keys together and returning the result directly */
+        void calculate_shared_public_key(
+            const std::vector<Crypto::PublicKey> &publicKeys,
+            Crypto::PublicKey &sharedPublicKey)
+        {
+            /* Remove any duplicate keys from the keys that are being
+               added together as we only use unique values */
+            std::vector<Crypto::PublicKey> keys = dedupeKeys(publicKeys);
+
+            /* Sort the keys in order as it is important that they
+               are added together in order */
+            sortKeys(keys);
+
+            sharedPublicKey = addKeys(keys);
+        }
+
+        /* Helper method that calculates a shared public key by adding the vector
+           of public keys together and returning the result directly */
+        Crypto::PublicKey calculate_shared_public_key(const std::vector<Crypto::PublicKey> &publicKeys)
+        {
+            Crypto::PublicKey publicKey;
+
+            calculate_shared_public_key(publicKeys, publicKey);
+
+            return publicKey;
+        }
+
+        /* Generates a partial signing key using the pre-prepared ring signature
+           and our pviate spend key and returns the resulant signing key that is
+           okay to distribute between the other multisig participants for
+           completing a ring signature */
+        Crypto::SecretKey
+            generate_partial_signing_key(const Crypto::Signature &signature, const Crypto::SecretKey &privateSpendKey)
+        {
+            /* This is kind of a hack because we we just need the
+               first 32-bytes of the existing signature that we will
+               "sign" via the sc_mul call below */
+            const Crypto::SecretKey signatureAsKey(signature.data);
+
+            Crypto::SecretKey result;
+
+            /* Multiplies the private key against the first 32-bytes of
+               the signature to come up with the scalar that is part
+               of the full signature for the given signature */
+            sc_mul(
+                reinterpret_cast<unsigned char *>(&result),
+                reinterpret_cast<const unsigned char *>(&signatureAsKey),
```

Comment:
```suggestion
                reinterpret_cast<const unsigned char *>(&signature),
```
Would just this not work, without making the tmp variable, since `sc_mul` is just going to read 32 bytes anyway?

Reply:
That’s obscure to read and isn’t documented behavior. I’d much rather have this very readable.